contracts/src/v0.8/dev/BatchBHS.sol

```diff
@@ -0,0 +1,61 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.6;
+
+/**
+ * @title BatchBHS
+ * @notice The BatchBHS contract acts as a proxy to write many blockhashes to the
+ *   provided BlockhashStore contract efficiently in a single transaction. This results
+ *   in plenty of gas savings and higher throughput of blockhash storage, which is desirable
+ *   in times of high network congestion.
+ */
+contract BatchBHS {
+  BlockhashStore public immutable BHS;
+
+  constructor(address blockhashStoreAddr) {
+    BHS = BlockhashStore(blockhashStoreAddr);
+  }
+
+  /**
+   * @notice stores blockhashes of the given block numbers in the configured blockhash store, assuming
+   *   they are availble though the blockhash() instruction.
+   * @param blockNumbers the block numbers to store the blockhashes of. Must be available via the
+   *   blockhash() instruction, otherwise this function call will revert.
+   */
+  function store(uint256[] memory blockNumbers) public {
+    for (uint256 i = 0; i < blockNumbers.length; i++) {
+      BHS.store(blockNumbers[i]);
+    }
+  }
+
+  /**
+   * @notice stores blockhashes after verifying blockheader of child/subsequent block
+   * @param blockNumbers the block numbers whose blockhashes should be stored, in decreasing order
+   * @param headers the rlp-encoded block headers of blockNumbers[i] + 1.
+   */
+  function storeVerifyHeader(uint256[] memory blockNumbers, bytes[] memory headers) public {
+    require(blockNumbers.length == headers.length, "input array arg lengths mismatch");
+    for (uint256 i = 0; i < blockNumbers.length; i++) {
+      BHS.storeVerifyHeader(blockNumbers[i], headers[i]);
+    }
+  }
+
+  /**
+   * @notice retrieves blockhashes of all the given block numbers from the blockhash store, if available.
+   * @param blockNumbers array of block numbers to fetch blockhashes for
+   */
+  function getBlockhashes(uint256[] memory blockNumbers) external view returns (bytes32[] memory) {
```

Comment:
Similar to the other comment, I think this one may not be that useful if it reverts if any are missing. The value would be knowing which ones are present / missing

Reply:
Yeah, actually I think you're right. Seems like we have a couple of options on how to do that here. 

We could simply have the entry in the returned array as `0x0` if the blockhash is missing (this is what the `blockhash` instruction returns if the block is too old), and the caller would have to match the index of the 0x0 in the returned array to the input array to get the block number that was missing.

We could additionally return a separate array with the block numbers that don't have an associated blockhash in the BHS, though that might be confusing as a function interface.

Reply:
In practice I feel like we only ever need to know which are missing / present, we could probably just return a list of block numbers for which the blockhashes aren't present?

If only BlockhashStore's s_blockhashes wasn't internal..would make life a lot easier haha

Reply:
Opted to return `0x0` in the returned array to indicate missing blockhash for the associated block number at that index.